karabo_data/detector_data.py

```diff
@@ -0,0 +1,393 @@
+            return None
+
+        group = key.partition('.')[0]
+        firsts, counts = file.get_index(source, group)
+        first, count = firsts[pos], counts[pos]
+
+        pulse_ids = (file.file['/INSTRUMENT/{}/{}/pulseId'.format(source, group)]
+                              [first:first + count])
+        # Raw files have a spurious extra dimension
+        if pulse_ids.ndim >= 2 and pulse_ids.shape[1] == 1:
+            pulse_ids = pulse_ids[:, 0]
+
+        if isinstance(self.pulses, by_id):
+            positions = self._select_pulse_ids(pulse_ids)
+        else:  # by_index
+            positions = self._select_pulse_indices(count)
+        pulse_ids = pulse_ids[positions]
+        train_ids = np.array([tid] * len(pulse_ids), dtype=np.uint64)
+        train_pulse_ids = pd.MultiIndex.from_arrays([train_ids, pulse_ids],
+                                                  names=['train', 'pulse'])
+
+        if isinstance(positions, slice):
+            data_positions = slice(
+                int(first + positions.start),
+                int(first + positions.stop),
+                positions.step
+            )
+        else:  # ndarray
+            # h5py fancy indexing needs a list, not an ndarray
+            data_positions = list(first + positions)
+
+        return _guess_axes(ds[data_positions], train_pulse_ids)
+
+    def _select_pulse_ids(self, pulse_ids):
+        """Select pulses by ID
+
+        Returns an array or slice of the indexes to include.
+        """
+        val = self.pulses.value
+        N = len(pulse_ids)
+        if isinstance(val, slice):
+            if val.step == 1:
+                after_start = np.nonzero(pulse_ids >= val.start)[0]
+                after_stop = np.nonzero(pulse_ids >= val.stop)[0]
+                start_ix = after_start[0] if (after_start.size > 0) else N
+                stop_ix = after_stop[0] if (after_stop.size > 0) else N
+                return slice(start_ix, stop_ix)
+
+            # step != 1
+            desired = np.arange(val.start, val.stop, step=val.step,
+                                dtype=np.uint64)
+
+        else:
+            desired = val
+
+        return np.nonzero(np.isin(pulse_ids, desired))[0]
+
+    def _select_pulse_indices(self, count):
+        """Select pulses by index
+
+        Returns an array or slice of the indexes to include.
+        """
+        val = self.pulses.value
+        if isinstance(val, slice):
+            return slice(val.start, min(val.stop, count), val.step)
+
+        # ndarray
+        return val[val < count]
+
+    def _assemble_data(self, tid):
+        key_module_arrays = {}
+
+        for modno, source in sorted(self.data.modno_to_source.items()):
+
+            for key in self.data.data._keys_for_source(source):
+                # At present, all the per-pulse data is stored in the 'image' key.
+                # If that changes, this check will need to change as well.
+
+                if key.startswith('image.'):
```

Comment:
I find it a bit too crafty to check key names to decide which data it per-pulse.
Maybe you can check if any value in `count` is `>1`?

Reply:
Everything I can think of is a bit hackish. E.g. you could have a detector recording only a single frame per train; this is still pulse-resolved data, but it wouldn't pass the `count > 1` test.

For now, the detector DAQ seems to write data in a fairly consistent way: with groups called detector, header, image, trailer, and all the pulse-resolved data in image. So this seems like the cleanest thing I can do.

The other possibility is to make the user specify it explicitly, e.g. by having separate methods `get_pulse_array` and `get_train_array`. That way it's not up to karabo_data to get the right one. But the meaning may not be obvious, and the error messages if you use the wrong method may be mysterious.

Reply:
In my mind: `count == 1` means train resolved, `count > 1` means pulse resolved. Is there technical issue to treat `image.data` as train resolved in the case it contains only one image per train?

Concerning consistency, for example FastCCD images are under: `data.image.pixels`

Reply:
I think it should use the same code for accessing `image.data` whether it has one frame per train or multiple. E.g. you want the data to have the same number of dimensions even if one of them is length 1.

This whole module only works for AGIPD & LPD so far. Maybe we should hold it up to try to provide a unified interface for detectors like FastCCD and Jungfrau as well.

Reply:
What are the difference between train and pulse data in this implementation? Is it only the extra axis for `pulseId`?
If yes, would it make sens to treat the train related data as a special case of the pulse related data, (with only 1 value on the pulse axis)? basically everything under INSTRUMENT group becomes the same kind of data. It is also how it is effectively stored in the file, there is not differentiation between `train` and `pulse` related data.

Reply:
The important difference in reading it is that the `image` group has a `pulseId` dataset, which tells us which pulse each frame corresponds to (with the ID that the detector knows, anyway). The extra machinery for reading pulse-resolved data is all about reading the pulse IDs, doing selections using them if necessary, and connecting them up to the data.

So it's not practical to read per-train data as a special case of per-pulse data; there are no pulse IDs to work with in that case.

Do you have paths to some good examples of FastCCD & Jungfrau data? Maybe if I look at that, I'll get some ideas about how to generalise the ideas effectively.

Reply:
sent you an email

Reply:
Looking at the different detectors' data reminds me that *detector* is a much broader term than the two megahertz-megapixel detectors that have been the focus of much of our attention so far.

I think the interface in this PR is particularly valuable for the big detectors, because the existing karabo_data interface is not very convenient for working with data across 16 different sources, nor for selecting and annotating data with pulse IDs. So far, most of the smaller detectors seem to be running at 10Hz, they don't appear to know about pulse IDs, and they mostly appear to record data under a single source (2 sources for Jungfrau).

So instead of generalising, maybe the answer for now is to make the naming more specific, so you explicitly call something like `agipd_interface` or `lpd_interface`. Then we can figure out how to handle other detectors when they are pushing the bounds of what can be done with karabo_data.

Reply:
Makes sense to me.